### PR TITLE
Scaffold: Add support for optional relations with string IDs

### PR DESCRIPTION
**Pull request by Tobbe**
Adds support for optional relations with string IDs to the scaffold generator. Does this by converting `''` to `null`

Fixes https://github.com/redwoodjs/redwood/issues/4297

### CODE REVIEW

**Comment by Tobbe** [comment]
@dthyresson suggested to add `setEmptyAsNull: true` as a special validation rule. I guess along the other two special cases we have `valueAsBoolean` and `valueAsJSON`.

I already had the solution I have here in place when I read his comment, so I left my solution in the code. But `setEmptyAsNull` (or maybe even just `emptyAsNull`) does look kind of nice too. So I'm open to changing the implementation here.

**Comment by Tobbe** [comment]
@jtoar looks like you added `valueAsBoolean` and `valueAsJSON` what do you think about keeping this PR the way it is vs changing to `emptyAsNull`?

**Comment by dthyresson** [comment]
Will something similar be needed for a SelectList?

```ts
interface CellSelect extends CellSuccessProps<FindSpots> {
  spotId: string
}

export const Success = ({ spots, spotId }: CellSelect) => {
  return (
    <div>
      <Label
        name="spotId"
        className="rw-label"
        errorClassName="rw-label rw-label-error"
      >
        Spot id
      </Label>
      <SelectField
        name="spotId"
        defaultValue={spotId}
        className="rw-input"
        errorClassName="rw-input rw-input-error"
      >
        <option key={'spot-empty'} value={null}>
          Pick a spot ...
        </option>
        {spots?.map((spot) => {
          return (
            <option key={spot.id} value={spot.id}>
              {spot.name}
            </option>
          )
        })}
      </SelectField>
      <FieldError name="spotId" className="rw-field-error" />
    </div>
  )
}
```

Am having trouble unsetting the value -- but editing seems ok.

**Comment by jtoar** [comment]
TL;DR: I have another suggestion: what we if detect fields that aren't `required` (`validation={ required: false }}`) and do the coercion from empty to `null` under the hood?

Just to recap, `valueAsBoolean` and `valueAsJSON` are custom Redwood extensions to the react-hook-form `register` API (specifically its second argument, `RegisterOptions`: https://react-hook-form.com/api/useform/register). The logic behind adding them was: 
1. to give users upgrading from redwood's in-house form coercion (which did cover these cases) to react-hook-form's (which doesn't) a way to do what they were doing before 
2. redwood uses graphql, so validation and coercion are really important (otherwise we get errors like this)

`setEmptyAsNull` definitely falls under the second point (I'm not sure if our in-house form coercion had any logic around `setEmptyAsNull`). But the main thing I'm worried about is that we're extending react-hook-forms API and it feels like we should do that as little as possible. That doesn't preclude doing it (I took that liberty), just that we should be a little more careful than usual about it!

Note that the naming convention was also copied from react-hook-form. They have `valueAsNumber` and `valueAsDate`. And `setValueAs` for everything else. So `setEmptyAsNull` may actually be closer to one of the naming patterns they have going on; I'd have to think more about it.

And lastly, note that there's still an outstanding forms issue that cjreimer tried to bring to our attention and I've completely failed him so far so if you're reading this I'm terribly sorry!

**Comment by Tobbe** [comment]
> Will something similar be needed for a SelectList?

Yes. But we never generate SelectList with the scaffold generator, right? So nothing to do there for this PR I believe. But if we add it for TextField, at least it might help the user to both see that it has to be done and also how to do it.

**Comment by Tobbe** [comment]
> Note that the naming convention was also copied from react-hook-form. They have `valueAsNumber` and `valueAsDate`. And `setValueAs` for everything else. So `setEmptyAsNull` may actually be closer to one of the naming patterns they have going on; I'd have to think more about it.

To me, this is the naming logic:
 - `setFooAs`: Takes a function that sets the value
 - `fooAs`: Boolean property
And with that logic it should be `emptyAsNull: true`

**Comment by Tobbe** [comment]
> @Tobbe TL;DR: I have another suggestion: what if we [...] do the coercion from empty to `null` under the hood?

Are you referring to extending the `register` api even further here, or is there another way to do it under the hood you were thinking of?

**Comment by Tobbe** [comment]
Cross-linking: https://github.com/redwoodjs/redwood/issues/3609

**Comment by Tobbe** [comment]
> Are you referring to extending the `register` api even further here, or is there another way to do it under the hood you were thinking of?

I figured it out I think. And I like it! Might also be able to handle @dthyresson `SelectField` case automatically too! I have it working for text fields, but haven't quite figured it out for selects yet. Will push an alternative PR later today

**Comment by Tobbe** [comment]
```
        <SelectField
          name="addressId"
          defaultValue={props.user?.addressId}
          className="rw-input"
          errorClassName="rw-input rw-input-error"
        >
          <option value={undefined}>None</option>
          <option value={null}>Null</option>
          <option value="">Empty</option>
          <option value="aaa">Address A</option>
          <option value="bbb">Address B</option>
          <option value="ccc">Address C</option>
          <option value="ddd">Address D</option>
        </SelectField>
```

For `value={undefined}` and `value={null}` r-h-f will use the text instead (so "None" and "Null") for the value sent to the submit function. I looked at overriding that behaviour for optional relational fields. It's possible (of course it is) but not sure if we want all that custom behavior or not.

**Comment by Tobbe** [comment]
New PR up: #4302 

I actually prefer that one. So if it's cool with you all I'd like to close this one and merge the other one.

**Comment by Tobbe** [comment]
Closing this in favor of #4302
